repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java

```diff
@@ -472,9 +483,11 @@ private boolean incrementAndCheckIfRelationsLimitReached(AtlasEdge atlasEdge, bo
             outLineageInfo.setHasMoreOutputs(true);
             hasRelationsLimitReached = true;
         } else {
-            if (! (direction.equals(AtlasLineageOnDemandInfo.LineageDirection.INPUT) && outGuid.equals(baseGuid))) {
-                outLineageInfo.incrementOutputRelationsCount();
-            }
+            outLineageInfo.incrementOutputRelationsCount();
+        }
+
+        if (lineageContainsVisitedEdge) {
```

Comment:
don't really need this if check. it is better if we update relationsOnDemand of ret in case of self-cyclic as well.